Mathlib/GroupTheory/CommutingProbability.lean

```diff
@@ -132,3 +135,83 @@ theorem inv_card_commutator_le_commProb : (↑(Nat.card (commutator G)))⁻¹ 
     (le_trans (ge_of_eq (commProb_eq_one_iff.mpr (Abelianization.commGroup G).mul_comm))
       (commutator G).commProb_quotient_le)
 #align inv_card_commutator_le_comm_prob inv_card_commutator_le_commProb
+
+-- Construction of group with commuting probability 1/n
+namespace DihedralGroup
+
+lemma commProb_odd {n : ℕ} (hn : Odd n) :
+    commProb (DihedralGroup n) = (n + 3) / (4 * n) := by
+  rw [commProb_def', DihedralGroup.card_conjClasses_odd hn, nat_card]
+  qify [show 2 ∣ n + 3 by rw [Nat.dvd_iff_mod_eq_zero, Nat.add_mod, Nat.odd_iff.mp hn]; rfl]
+  rw [div_div, ← mul_assoc]
+  congr
+
+lemma div_two_lt {n : ℕ} (h0 : n ≠ 0) : n / 2 < n :=
+  Nat.div_lt_self (Nat.pos_of_ne_zero h0) (lt_add_one 1)
+
+lemma div_four_lt : {n : ℕ} → (h0 : n ≠ 0) → (h1 : n ≠ 1) → n / 4 + 1 < n
+| 0 | 1 | 2 | 3 => by decide
+| n + 4 => by intros; linarith [n.add_div_right four_pos, n.div_le_self 4]
+
+/-- A list of Dihedral groups whose product will have commuting probability `1 / n`. -/
+def reciprocalFactors (n : ℕ) : List ℕ :=
+  if h0 : n = 0 then [0]
+  else if h1 : n = 1 then []
+  else if 2 ∣ n then
+    have := div_two_lt h0
+    3 :: reciprocalFactors (n / 2)
+  else
+    have := div_four_lt h0 h1
+    n % 4 * n :: reciprocalFactors (n / 4 + 1)
+
+/-- A finite product of Dihedral groups. -/
+def Product (l : List ℕ) : Type :=
+  ∀ i : Fin l.length, DihedralGroup l[i]
+
+instance (l : List ℕ) : Group (Product l) := Pi.group
+
+lemma commProb_nil : commProb (Product []) = 1 := by
+  simp [Product, commProb_pi]
+
+lemma commProb_cons (n : ℕ) (l : List ℕ) :
+    commProb (Product (n :: l)) = commProb (DihedralGroup n) * commProb (Product l) := by
+  simp [Product, commProb_pi, Fin.prod_univ_succ]
+
+lemma commProb_product (l : List ℕ) :
+    commProb (Product l) = (l.map (fun k => commProb (DihedralGroup k))).prod := by
+  induction' l with n l h
+  · rw [commProb_nil, List.map_nil, List.prod_nil]
+  · rw [commProb_cons, List.map_cons, List.prod_cons, h]
+
+/-- Construction of a group with commuting probability `1 / n`. -/
+theorem commProb_reciprocal (n : ℕ) :
+    commProb (Product (reciprocalFactors n)) = 1 / n := by
+  rw [commProb_product]
+  unfold reciprocalFactors
```

Comment:
Could you write some API about this definition so that it doesn't need unfolding here?